tcmur_cmd_handler.c

```diff
@@ -144,6 +144,124 @@ static int write_work_fn(struct tcmu_device *dev,
+	uint64_t end_lba;
+	uint32_t max_lbas, copy_lbas;
+	uint32_t max_sectors;
+	int ret;
+
+	end_lba = start_lba + lba_cnt;
+
+	max_sectors = tcmu_get_attribute(dev, "hw_max_sectors");
+	max_lbas = min(max_sectors, ((uint32_t)(~0U)));
+
+	tcmu_dbg("Start lba: %llu Number of lba:: %hu, end lba: %llu\n",
+		 start_lba, lba_cnt, end_lba);
+
+	/*
+	 * Write contents of the logical block data(from the Data-Out Buffer)
+	 * to each LBA in the specified LBA range.
+	 */
+	while(start_lba < end_lba) {
+		copy_lbas = min(max_lbas, lba_cnt);
+
+		if (start_lba + copy_lbas >= end_lba)
+			cmd->done = handle_generic_cbk;
+
+		ret = rhandler->write(dev, cmd, cmd->iovec, cmd->iov_cnt,
+				      tcmu_iovec_length(cmd->iovec, cmd->iov_cnt),
+				      block_size * copy_lbas);
+		if (ret)
+			goto out;
+
+		start_lba += copy_lbas;
+		lba_cnt -= copy_lbas;
+	}
+
+out:
+	return ret;
+}
+
+static int handle_write_same_16(struct tcmu_device *dev, struct tcmulib_cmd *cmd)
+{
+	uint8_t *cdb = cmd->cdb;
+	uint8_t *sense = cmd->sense_buf;
+	uint32_t blocks = tcmu_get_xfer_length(cdb);
+	uint32_t block_size = tcmu_get_dev_block_size(dev);
+	uint64_t start_lba = tcmu_get_lba(cmd->cdb);
+	uint64_t dev_last_lba = tcmu_get_dev_num_lbas(dev);
+
+	if (cmd->iov_cnt != 1 || cmd->iovec->iov_len != block_size) {
+		tcmu_err("Illegal Data-Out: iov_cnt %u length: %u\n",
+			 cmd->iov_cnt, cmd->iovec->iov_len);
+		return tcmu_set_sense_data(sense, ILLEGAL_REQUEST,
+					   ASC_INVALID_FIELD_IN_CDB,
+					   NULL);
+	}
+
+	/*
+	 * From sbc4r13, section 5.50 WRITE SAME (16) command
+	 *
+	 * A write same (WSNZ) bit has beed set to one, so the device server
+	 * won't support a value of zero here.
+	 */
+	if (!blocks) {
+		tcmu_err("The WSNZ = 1 & WRITE_SAME blocks = 0 is not supported!\n");
+		return tcmu_set_sense_data(sense, ILLEGAL_REQUEST,
+					   ASC_INVALID_FIELD_IN_CDB,
+					   NULL);
+	}
+
+	/*
+	 * The MAXIMUM WRITE SAME LENGTH field in Block Limits VPD page (B0h)
+	 * limit the maximum block number for the WRITE SAME.
+	 */
+	if (blocks > VPD_MAX_WRITE_SAME_LENGTH) {
+		tcmu_err("blocks: %u exceeds MAXIMUM WRITE SAME LENGTH: %u\n",
+			 blocks, VPD_MAX_WRITE_SAME_LENGTH);
+		return tcmu_set_sense_data(sense, ILLEGAL_REQUEST,
+					   ASC_INVALID_FIELD_IN_CDB,
+					   NULL);
+	}
+
+	/*
+	 * The logical block address plus the number of blocks shouldn't
+	 * exceeds the capacity of the medium
+	 */
+	if (start_lba + blocks > dev_last_lba) {
+		tcmu_err("lba %llu plus blocks %lu exceeds the dev capacity %llu\n",
+			 start_lba, blocks, dev_last_lba);
+		return tcmu_set_sense_data(sense, ILLEGAL_REQUEST,
+					   ASC_LBA_OUT_OF_RANGE,
+					   NULL);
+	}
+
+	/*
+	 * For now not support UNMAP bit and ANCHOR bit.
+	 */
+	if (cdb[1] & 0x10 || cdb[1] & 0x08) {
+		tcmu_err("The ANCHOR or UNMAP bit is set not support!\n");
+		return tcmu_set_sense_data(sense, ILLEGAL_REQUEST,
+					   ASC_INVALID_FIELD_IN_CDB,
+					   NULL);
+	}
+
+	cmd->done = handle_write_same_16_null_cbk;
```

Comment:
Just a question, but does WRITE_SAME zero work with qcow implemented this way? I am not sure....